File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/promises/extensions/LoopedInPromise.java
package com.lasthopesoftware.bluewater.shared.promises.extensions;

import android.content.Context;
import android.os.Handler;

import com.namehillsoftware.handoff.Messenger;
import com.namehillsoftware.handoff.promises.MessengerOperator;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.queued.MessageWriter;
import com.namehillsoftware.handoff.promises.queued.PreparedMessengerOperator;
import com.namehillsoftware.handoff.promises.queued.cancellation.CancellableMessageWriter;
import com.namehillsoftware.handoff.promises.queued.cancellation.CancellablePreparedMessengerOperator;
import com.namehillsoftware.handoff.promises.response.ImmediateResponse;
import com.namehillsoftware.handoff.promises.response.PromisedResponse;

import org.joda.time.Duration;

public class LoopedInPromise<Result> extends Promise<Result> {

	public LoopedInPromise(MessageWriter<Result> task, Context context) {
		this(task, new Handler(context.getMainLooper()));
	}

	public LoopedInPromise(MessageWriter<Result> task, Handler handler) {
		super(new Executors.LoopedInResponse<>(new PreparedMessengerOperator<>(task), handler));
	}

	public LoopedInPromise(MessageWriter<Result> task, Context context, Duration delay) {
		this(task, new Handler(context.getMainLooper()), delay);
	}

	public LoopedInPromise(MessageWriter<Result> task, Handler handler, Duration delay) {
		super(new Executors.DelayedLoopedInPromise<>(new PreparedMessengerOperator<>(task), handler, delay));
	}

	public LoopedInPromise(CancellableMessageWriter<Result> task, Handler handler) {
		super(new Executors.LoopedInResponse<>(new CancellablePreparedMessengerOperator<>(task), handler));
	}

	public LoopedInPromise(MessengerOperator<Result> task, Handler handler) {
		super(new Executors.LoopedInResponse<>(task, handler));
	}

	public static <TResult, TNewResult> PromisedResponse<TResult, TNewResult> response(ImmediateResponse<TResult, TNewResult> task, Context context) {
		return response(task, new Handler(context.getMainLooper()));
	}

	public static <TResult, TNewResult> PromisedResponse<TResult, TNewResult> response(ImmediateResponse<TResult, TNewResult> task, Handler handler) {
		return new OneParameterExecutors.ReducingFunction<>(task, handler);
	}

	private static class Executors {
		static final class LoopedInResponse<Result> implements MessengerOperator<Result>, Runnable {

			private final MessengerOperator<Result> task;
			private final Handler handler;
			private Messenger<Result> resultMessenger;

			LoopedInResponse(MessengerOperator<Result> task, Handler handler) {
				this.task = task;
				this.handler = handler;
			}

			@Override
			public void send(Messenger<Result> resultMessenger) {
				this.resultMessenger = resultMessenger;

				if (handler.getLooper().getThread() == Thread.currentThread())
					run();
				else
					handler.post(this);
			}

			@Override
			public void run() {
				task.send(resultMessenger);
			}
		}

		static final class DelayedLoopedInPromise<Result> implements MessengerOperator<Result>, Runnable {
			private final MessengerOperator<Result> task;
			private final Handler handler;
			private final Duration delay;
			private Messenger<Result> resultMessenger;

			DelayedLoopedInPromise(MessengerOperator<Result> task, Handler handler, Duration delay) {
				this.task = task;
				this.handler = handler;
				this.delay = delay;
			}

			@Override
			public void send(Messenger<Result> resultMessenger) {
				this.resultMessenger = resultMessenger;
				handler.postDelayed(this, (int)delay.getMillis());
			}

			@Override
			public void run() {
				task.send(resultMessenger);
			}
		}
	}

	private static class OneParameterExecutors {

		static class ReducingFunction<TResult, TNewResult> implements PromisedResponse<TResult, TNewResult>, MessageWriter<TNewResult> {

			private final ImmediateResponse<TResult, TNewResult> callable;
			private final Handler handler;

			private TResult result;

			ReducingFunction(ImmediateResponse<TResult, TNewResult> callable, Handler handler) {
				this.callable = callable;
				this.handler = handler;
			}

			@Override
			public TNewResult prepareMessage() throws Throwable {
				return callable.respond(result);
			}

			@Override
			public Promise<TNewResult> promiseResponse(TResult result) {
				this.result = result;
				return new LoopedInPromise<>(this, handler);
			}
		}
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/promises/extensions/LoopedInPromise.kt
package com.lasthopesoftware.bluewater.shared.promises.extensions

import android.content.Context
import android.os.Handler
import com.namehillsoftware.handoff.Messenger
import com.namehillsoftware.handoff.promises.MessengerOperator
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.handoff.promises.queued.MessageWriter
import com.namehillsoftware.handoff.promises.queued.PreparedMessengerOperator
import com.namehillsoftware.handoff.promises.queued.cancellation.CancellableMessageWriter
import com.namehillsoftware.handoff.promises.queued.cancellation.CancellablePreparedMessengerOperator
import com.namehillsoftware.handoff.promises.response.ImmediateResponse
import com.namehillsoftware.handoff.promises.response.PromisedResponse
import org.joda.time.Duration

open class LoopedInPromise<Result> : Promise<Result> {
	constructor(task: MessageWriter<Result>, context: Context)
		: this(task, Handler(context.mainLooper))

	constructor(task: MessageWriter<Result>, handler: Handler)
		: super(Executors.LoopedInResponse<Result>(PreparedMessengerOperator<Result>(task), handler))

	constructor(task: MessageWriter<Result>, context: Context, delay: Duration)
		: this(task, Handler(context.mainLooper), delay)

	constructor(task: MessageWriter<Result>, handler: Handler, delay: Duration)
		: super(Executors.DelayedLoopedInPromise<Result>(PreparedMessengerOperator<Result>(task), handler, delay))

	constructor(task: CancellableMessageWriter<Result>, handler: Handler)
		: super(Executors.LoopedInResponse<Result>(CancellablePreparedMessengerOperator<Result>(task), handler))

	constructor(task: MessengerOperator<Result>, handler: Handler)
		: super(Executors.LoopedInResponse<Result>(task, handler))

	private class Executors {
		class LoopedInResponse<Result>(private val task: MessengerOperator<Result>, private val handler: Handler) : MessengerOperator<Result>, Runnable {
			private lateinit var resultMessenger: Messenger<Result>

			override fun send(resultMessenger: Messenger<Result>) {
				this.resultMessenger = resultMessenger
				if (handler.looper.thread === Thread.currentThread()) run() else handler.post(this)
			}

			override fun run() {
				try {
					task.send(resultMessenger)
				} catch (err: Throwable) {
					resultMessenger.sendRejection(err)
				}
			}
		}

		class DelayedLoopedInPromise<Result>(private val task: MessengerOperator<Result>, private val handler: Handler, private val delay: Duration) : MessengerOperator<Result>, Runnable {
			private lateinit var resultMessenger: Messenger<Result>

			override fun send(resultMessenger: Messenger<Result>) {
				this.resultMessenger = resultMessenger
				handler.postDelayed(this, delay.millis)
			}

			override fun run() {
				try {
					task.send(resultMessenger)
				} catch (err: Throwable) {
					resultMessenger.sendRejection(err)
				}
			}
		}
	}

	private class OneParameterExecutors {
		class ReducingFunction<TResult, TNewResult>(private val callable: ImmediateResponse<TResult, TNewResult>,	private val handler: Handler) :
			PromisedResponse<TResult, TNewResult>, MessageWriter<TNewResult> {

			private var result: TResult? = null

			override fun prepareMessage(): TNewResult = callable.respond(result)

			override fun promiseResponse(result: TResult): Promise<TNewResult> {
				this.result = result
				return LoopedInPromise(this, handler)
			}
		}
	}

	companion object {
		@JvmStatic
		fun <TResult, TNewResult> response(task: ImmediateResponse<TResult, TNewResult>, context: Context): PromisedResponse<TResult, TNewResult> {
			return response(task, Handler(context.mainLooper))
		}

		@JvmStatic
		fun <TResult, TNewResult> response(task: ImmediateResponse<TResult, TNewResult>, handler: Handler): PromisedResponse<TResult, TNewResult> {
			return OneParameterExecutors.ReducingFunction(task, handler)
		}
	}
}
